content/knowledge_base/flare/consensus/avalanche/specification.md

```diff
@@ -0,0 +1,210 @@
+If a virtuous transaction fails to get accepted due to a problem with parents, it could be accepted if reissued with different parents.
+
+### Transaction Entangling
+
+How Avalance entangles transactions:
+
+```text
+init:
+    `Set T` = Ø set of known transactions (empty set)
+    Q = Ø set of queried transactions (empty set)
+
+generateTx:
+    edges = {T' <— T: T' ∈ parentSelection(`Set T`)}
+    T = Tx(data, edges)
+    onReceiveEvent(T)
+
+onReceiveEvent(t)
+    if T ∉ `Set T` then
+        if PT (conflict set of T) = Ø then
+            PT = {T}, Pt.pref = T
+            PT.last = T, PT.cnt = 0
+        else PT = PT U {T}
+        `Set T` = T U {T}, cT = 0
+```
+
+Because transactions that generate and consume the same UTXO do not conflict with each other, any transaction can be reissued with different parents.
+
+### Node Main Loop
+
+Protocol main loop executed by each node:
+
+```text
+AvalancheLoop:
+    while true do
+        find T that satisfies T ∈ `Set T` AND T ∉ Q
+        K = sample(N\u, k)
+        P = SUM(v ∈ k) query(v, T)
+
+        if P ≥ α then
+            cT = 1 (chit for transaction T)
+            update preferences for ancestors:
+            for T' ∈ `Set T`: T' *<— T do
+                if T' confidence is larger than the confidence of preferred transation T` in the P conflict set - update the prefered transaction of the set:
+                if d(T') > d(PT'.pref) then
+                    PT'.pref = T'
+                if T' is not the last transaction in the T' conflict set - update the last place with it
+                if T' ≠ PT'.last then
+                    PT'.last = T', PT'.cnt=1
+                else
+                    ++PT'.cnt
+        else
+            for T' ∈ `Set T`: T' *<— T do
+                PT'.cnt = 0
+
+            otherwise confidence cnt remains 0 forever
+            Q = Q U {T} - mark T as queried
+```
+
+In each iteration the node tries to select a transaction that has not yet been queried.
+If no such transaction exists, the loop will stall until a new transation is added to the Set T.
+The node then select k number of peers and queries them.
+If more than α of these peers return a positive response, the chit value is set to 1.
+After this tep the loop updates the prefered transaction of each conflict set of the transactions in its ancestry.
+Next, T is added to the set Q of queried transactions so it will never be queried again by the node.
+
+### Query for Transaction
+
+What happens when a node receives a query for transaction T from peer j:
+
+```text
+isPreferred(T)
+    return T = PT.pref
+
+isStronglyPreferred(T)
+    return ∀T' ∈ `Set T`, T' *<— T : isPreferred(T')
+
+isAccepted(T)
+    return
+        ((∀T' ∈ `Set T`, T' <— T: isAccepted(T'))
+            ∩ |PT|=1 ∩ PT.cnt ≥ β1)  safe early commitment
+        ∪ (PT.cnt ≥ β2)              consecutive counter
+
+onQuery(j, T)
+    onReceiveTx(T)
+    respond(, isStronglyPreferred(T))
```

Comment:
In other code blocks, functions were defined with the keyword `procedure` and followed by a semicolon. Why is it not the case here? I'm very confused as to what language is used to describe those algorithms haha

Reply:
Not sure what you mean, I have used 'procedure' anywhere.

Reply:
Semicolon were missing because I missed them here, added https://github.com/optakt/optakt.github.io/pull/13/commits/993d3f700e85aaa7570bdbd31106fc78b8090933#diff-2d4f16e70781857aec6689a748ad155e45611b81f184aee076bb139890410b98R171
Thanks!

Reply:
My bad, I was confusing it with the Avalanche whitepaper pseudocode :p It's all good as it is now then for me :)